lua/tardis/libraries/sh_settings.lua

```diff
@@ -80,66 +49,236 @@ function TARDIS:GetSettingsData()
+-- Accessing
+
+function TARDIS:SetSetting(id, value)
+    local data = self.SettingsData[id]
+    if not data then error("Requested setting ", id, " does not exist") end
+
+    if value ~= nil and data.type == "integer" then
+        value = math.Round(value)
+    end
+
+    if value ~= nil and (data.type == "number" or data.type == "integer") then
+        if data.min and data.max then
+            value = math.max(data.min, math.min(data.max, value))
+        end
+        if data.round_func then
+            value = data.round_func(value)
+        end
+    end
+
+    if value == nil and data.class == "global" then
+        value = data.value
+    end
+
     if SERVER then
-        self.Settings[id]=value
-    elseif networked then
-        self.NetworkedSettings[id]=value
+        if data.class == "global" then
+            self.GlobalSettings[id]=value
+
+            if data.convar then
+                local convar = GetConVar(data.convar.name)
+                if data.type == "integer" then
+                    convar:SetInt(value)
+                elseif data.type == "number" then
+                    convar:SetFloat(value)
+                elseif data.type == "bool" then
+                    convar:SetBool(value)
+                else
+                    convar:SetString(tostring(value))
+                end
+            end
+        else
+            error("Setting " .. id .. " is being set serverside, but is not global")
+        end
     else
-        self.LocalSettings[id]=value
+        if data.class == "networked" then
+            self.NetworkedSettings[id]=value
+        elseif data.class == "local" then
+            self.LocalSettings[id]=value
+        elseif data.class == "global" then
+            TARDIS:GlobalSettingChange(id, value)
+        else
+            error("Setting " .. id .. " is being set clientside, but has unsupported class")
+        end
     end
+
     self:SaveSettings()
-    if (SERVER and broadcast) or networked then
-        self:SendSetting(id,value)
+
+    if (SERVER and data.class == "global") or (CLIENT and data.class == "networked") then
+        self:SendSetting(id, value)
+
+        for k,v in pairs(ents.FindByClass("gmod_tardis")) do
+            v:CallCommonHook("GlobalSettingChanged", id, value)
+        end
     end
+
+    return value
 end
 
-function TARDIS:GetSetting(id,default,ply)
-    if ply and IsValid(ply) and self.ClientSettings[ply:UserID()] and (self.ClientSettings[ply:UserID()][id]~=nil) then
-        return self.ClientSettings[ply:UserID()][id]
-    elseif self.Settings[id] ~= nil then
-        return self.Settings[id]
-    elseif CLIENT and (self.LocalSettings[id] ~= nil) then
-        return self.LocalSettings[id]
-    elseif CLIENT and (self.NetworkedSettings[id]~=nil) then
-        return self.NetworkedSettings[id]
-    else
-        return default
+function TARDIS:GetSetting(id, src)
+    local ply
+    if IsEntity(src) and src:IsPlayer() then
```

Comment:
This can be slightly optimised to only call IsEntity once e.g.

```lua
if IsEntity(src) then
  if src:IsPlayer() then
     ...
   else
     ...
   end
end
```

This could affect performance as settings can be called every frame multiple times

Reply:
Sure, fixed